### PR TITLE
Extract expect/actual structure for core+generated types

### PR DESCRIPTION
**Pull request by JustinMullin**
Extract base structure for wrapper library to a common module using `expect` declarations, and refactor native implementations to use `actual`. Move shared methods on core types to extension functions to reduce duplication between platform-specific implementations.

### CODE REVIEW

**Comment by JustinMullin** [comment]
Opening this to preview what utilizing the MPP `expect`/`actual` keywords would look like for this project, as this is likely to be a sticking point in implementing https://github.com/utopia-rise/godot-kotlin/issues/10. There's some overhead introduced here, and a couple of design considerations.

If it's decided that godot-kotlin is moving this direction, it would be good to get the refactoring out of the way, as it does involve some pretty significant shifting of code and this branch will drift rapidly as modifications are made to core classes. I also recognize that the JVM platform implementation is a lower priority, so don't want to push too hard on this path if there are concerns.

A few highlights/potential places where this may be controversial:
- The `expect`ed class can't contain default implementations, so Godot-external logic in the core classes needs to move to extensions on the core class (the route demonstrated in this PR), or otherwise be duplicated in each platform-specific implementation.
- With the extensions route, there are some ramifications for imports, as each extension method is an individual import. As a result, many imports may be needed in some cases, for users that are disinclined to use the wildcard import `import godot.core.*`
- An `expect` class can't define `const` `val`s. I'm not sure if this is a restriction that will be lifted at a later date by K/N, but it's true at present. Here I've removed the `const` modifier from `val`s that manifest in the core/generated classes; I don't believe this should cause any problems as inlining wasn't likely gaining much in these cases, but it could be a future consideration.

**Comment by piiertho** [comment]
Hi ! 
Great work !
I cannot build godot-library with parameter platform, using:
`./gradlew :wrapper:godot-library:build -Pplatform=macos`
(It does metadata but no compilation)

**Comment by JustinMullin** [comment]
Hey @piiertho - do you mean you're seeing the `godot-library-metadata` artifact published but not the `godot-library-macos` artifact? I'm able to build the `godot-library-macos` artifacts using that command locally; cleaned the Gradle project and cleared Maven local artifacts, and was able to build+run the sample games from scratch. The CI check seems to be successfully running `:wrapper:godot-library:compileKotlinMacos` too here. It's entirely possible I'm missing something cached somewhere locally though.

**Comment by piiertho** [comment]
Oh sorry my bad, was trying to compile Windows from the wsl commandline

**Comment by CedNaru** [comment]
Impressive Job.

Is there some overhead when using extension fonctions instead of regular methods ? 
Otherwise I see no issue with using them. There is no private properties needed in the common code so I don't think we will encounter any visibily issue

**Comment by JustinMullin** [comment]
Nope, no overhead to extension functions (other than potentially a minimal compiler cost). They end up as static methods with the receiver as a parameter in the bytecode, so just a regular method invocation.

File: samples/games/kotlin/gradle.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+platform=macos
```
Comment (piiertho) [inline samples/games/kotlin/gradle.properties]: seems to have commit a wrong file :)

**Comment by piiertho** [comment]
@JustinMullin you put this PR as a draft, do you want to continue working on it or should we review it ?

File: samples/games/kotlin/gradle.properties (hunk 1 above)
Comment (JustinMullin) [reply samples/games/kotlin/gradle.properties]: Whoops! Deleted now, good catch.

**Comment by JustinMullin** [comment]
> @JustinMullin you put this PR as a draft, do you want to continue working on it or should we review it ?

I opened as draft expecting this might initially just generate conversation vs. be immediately merged, but it seems it's generally pretty accepted as an approach here, so I'll mark ready for review to kick that off formally. =)

**Comment by chippmann** [comment]
@JustinMullin Really great work! Thanks a lot for your contribution!
@piiertho I wouldn't merge it though. I would first merge the entry generation branch once it's finished, so we can make a release first. After that, we can merge this i think as after that this will be the biggest task.
Are you all ok with that?

**Comment by piiertho** [comment]
@chippmann I'm ok with that, I will complete Android and iOS compatibility too, before making a release.

**Comment by chippmann** [comment]
@JustinMullin If you are ok with it I'll close this PR as it becomes obsolete and incomaptible with the merge of the `master-merge` branch. As said we can discuss the whole JVM and expect/actual topic once the new binding is in place :-)

**Comment by chippmann** [comment]
Closing in favor of `master-merge`

**Pull request closed without merging**